src/server/mysql_prepared_statement.ipp

```diff
@@ -0,0 +1,161 @@
+/*
+   Copyright (C) 2016 by Sergey Popov <loonycyborg@gmail.com>
+   Part of the Battle for Wesnoth Project http://www.wesnoth.org
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License 2
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+
+   See the COPYING file for more details.
+*/
+
+#ifndef MYSQL_PREPARED_STATEMENT_IPP
+#define MYSQL_PREPARED_STATEMENT_IPP
+
+#include <array>
+#include <utility>
+#include <string>
+#include <string.h>
+#include <iostream>
+
+#include <mysql/mysql.h>
+
+#include "exceptions.hpp"
+
+struct sql_error : public game::error
+{
+	sql_error(const std::string& message) : game::error(message) {}
+};
+
+MYSQL_BIND make_bind(const std::string& str, my_bool* is_null = 0)
+{
+	MYSQL_BIND result;
+	memset(&result, 0, sizeof (MYSQL_BIND));
+	result.buffer_type = MYSQL_TYPE_STRING;
+	result.buffer = const_cast<void*>(static_cast<const void*>(str.c_str()));
+	result.buffer_length = str.size();
+	result.is_unsigned = 0;
+	result.is_null = is_null;
+	result.length = 0;
+	return result;
+}
+
+MYSQL_BIND make_bind(char* str, std::size_t* len, my_bool* is_null = 0)
+{
+	MYSQL_BIND result;
+	memset(&result, 0, sizeof (MYSQL_BIND));
+	result.buffer_type = MYSQL_TYPE_STRING;
+	result.buffer = static_cast<void*>(str);
+	result.buffer_length = *len;
+	result.is_unsigned = 0;
+	result.is_null = is_null;
+	result.length = len;
+	return result;
+}
+
+MYSQL_BIND make_bind(int& i, my_bool* is_null = 0)
+{
+	MYSQL_BIND result;
+	memset(&result, 0, sizeof (MYSQL_BIND));
+	result.buffer_type = MYSQL_TYPE_LONG;
+	result.buffer = static_cast<void*>(&i);
+	result.is_unsigned = 0;
+	result.is_null = is_null;
+	return result;
+}
+
+MYSQL_BIND make_bind(int&& i, my_bool* is_null = 0)
+{
+	MYSQL_BIND result;
+	memset(&result, 0, sizeof (MYSQL_BIND));
+	result.buffer_type = MYSQL_TYPE_LONG;
+	result.buffer = static_cast<void*>(&i);
```

Comment:
That's true, but the `int&` overload should handle that, not the `int&&` one, so I don't see a reason for the rvalue-ref one.

Won't the rvalue-ref overload only be picked over the regular ref if you use a temporary or std::move or the like? That's the very case in which it is unsafe.